### ISSUE TITLE
Array types cannot change nullability when generating DDL

### ISSUE THREAD

**Issue by mjburghoffer** [jOOQ/jOOQ#7811]
### Expected behavior and actual behavior:
I would like to be able to generate an explicity nullable or non-nullable array type using `DSL.ddl(Table)` method.  This is currently impossible due to the fact that calling `

### Steps to reproduce the problem (include example code if possible):
Create a custom table with a field created like:

```java
public final TableField<MyRecord, String[]> REGIONS = createField("regions", org.jooq.impl.SQLDataType.CLOB.getArrayDataType().nullable(false), this, "");
```

and render it like

```java
System.out.println(DSL.using(SQLDialect.POSTGRES).ddl(MyTable.MY_TABLE));
```

and the output will include a DDL statement like:

```sql
create table "objects"."test"(
  -- ...other columns...
  "regions" clob not null
  );
```

but would expect

```sql
create table "objects"."test"(
  -- ...other columns...
  "regions" text[] not null
  );
```

### Cause

`org.jooq.impl.ArrayDataType` does not override `DefaultDataType.nullability(Nullability n)` but should, and as such an array gets converted into a non-array when attempting to change its nullability.

```java
    @Override
    public final DataType<T> nullability(Nullability n) {
        return new DefaultDataType<T>(this, precision, scale, length, n, collation, n.nullable() ? false : identity, defaultValue);
    }
```

### Versions:

- jOOQ: 3.11.4
- Java: zulu-10
- Database (include vendor): postgres 9.6
- OS: mac osx

**Comment by lukaseder** [comment]
Thank you very much for your report. Will look into this shortly

**Comment by lukaseder** [comment]
I can reproduce this issue

**Comment by lukaseder** [comment]
This doesn't just affect nullability, but all other flags as well

**Comment by lukaseder** [comment]
Ideally, there would be no subtype of `DefaultDataType` - it only makes things like this much harder. I will need to think about this more thoroughly. There's a signficiant risk for regressions in that area.

**Comment by mjburghoffer** [comment]
Yeah this is a tough one, there really isn't any built-in strategies for doing type-safe clones in java.  I am reminded of how useful Kotlin data class `copy` methods are.  I wonder if this is solvable with code-generation or delegation?

**Comment by lukaseder** [comment]
@mjburghoffer It is probably very easy to solve, but hard to avoid regressions.

**Comment by rbuehlma** [comment]
@lukaseder We are facing this same issue with Jooq version 3.11.11. Is it possible to fix it in 3.11 or do you have a timeline when 3.12 will be released?

**Comment by lukaseder** [comment]
It seems that this could be fixed more easily than I feared. Rescheduling for 3.12.

@rbuehlma early Q3 2019 is the current plan. The roadmap for 3.12 is almost empty. We'll be ready soon.

**Comment by lukaseder** [comment]
Works indeed. Fixed for jOOQ 3.12
